# Financial Sentiment Analysis: Comparative Study of Naive Bayes and BERT

## Introduction
Financial sentiment analysis provides insights into market sentiments, helping investors and analysts make informed decisions. This project focuses on analyzing sentiments in financial reports using text classification methods, comparing the performance of traditional statistical model (Naive Bayes) and modern deep learning model (BERT models).

## Methodology:
- The data was preprocessed using tokenization, lemmatization, and vectorization.
- The Naive Bayes model was implemented using TF-IDF vectors, and the BERT model was fine-tuned on the financial sentiment dataset.

### Install libraries

In [1]:
pip install pandas transformers


  Using cached pytz-2024.1-py2.py3-none-any.whl.metadata (22 kB)
  Using cached tzdata-2024.1-py2.py3-none-any.whl.metadata (1.4 kB)
   ---------------------------------------- 0.0/11.5 MB ? eta -:--:--
   ------- -------------------------------- 2.1/11.5 MB 45.1 MB/s eta 0:00:01
   ---------------- ----------------------- 4.7/11.5 MB 50.3 MB/s eta 0:00:01
   ------------------------- -------------- 7.2/11.5 MB 51.3 MB/s eta 0:00:01
   ---------------------------------- ----- 9.8/11.5 MB 52.1 MB/s eta 0:00:01
   ---------------------------------------  11.5/11.5 MB 54.4 MB/s eta 0:00:01
   ---------------------------------------  11.5/11.5 MB 54.4 MB/s eta 0:00:01
   ---------------------------------------  11.5/11.5 MB 54.4 MB/s eta 0:00:01
   ---------------------------------------- 11.5/11.5 MB 31.2 MB/s eta 0:00:00
Using cached pytz-2024.1-py2.py3-none-any.whl (505 kB)
Using cached tzdata-2024.1-py2.py3-none-any.whl (345 kB)
Note: you may need to restart the kernel to use updated p

In [2]:
pip install IProgress TensorFlow torch torchvision torchaudio


  Using cached absl_py-2.1.0-py3-none-any.whl.metadata (2.3 kB)
  Using cached rich-13.7.1-py3-none-any.whl.metadata (18 kB)
     ---------------------------------------- 0.0/48.7 kB ? eta -:--:--
     ---------------------------------------- 48.7/48.7 kB ? eta 0:00:00
  Using cached markdown_it_py-3.0.0-py3-none-any.whl.metadata (6.9 kB)
  Using cached mdurl-0.1.2-py3-none-any.whl.metadata (1.6 kB)
   ---------------------------------------- 0.0/385.2 MB ? eta -:--:--
   ---------------------------------------- 1.8/385.2 MB 59.1 MB/s eta 0:00:07
   ---------------------------------------- 4.4/385.2 MB 56.5 MB/s eta 0:00:07
    --------------------------------------- 7.0/385.2 MB 55.6 MB/s eta 0:00:07
    --------------------------------------- 9.5/385.2 MB 55.4 MB/s eta 0:00:07
   - -------------------------------------- 12.0/385.2 MB 54.4 MB/s eta 0:00:07
   - -------------------------------------- 14.5/385.2 MB 54.4 MB/s eta 0:00:07
   - -------------------------------------- 17.1/3

In [3]:
!pip3 install torch torchvision torchaudio setuptools 
# !pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu118

In [4]:
pip install setuptools nltk

  Using cached nltk-3.8.1-py3-none-any.whl.metadata (2.8 kB)
  Using cached click-8.1.7-py3-none-any.whl.metadata (3.0 kB)
Using cached nltk-3.8.1-py3-none-any.whl (1.5 MB)
Using cached click-8.1.7-py3-none-any.whl (97 kB)
Note: you may need to restart the kernel to use updated packages.


## Data Loading and Preprocessing

In [5]:
import torch
print(torch.__version__)

2.3.1+cpu


In [6]:
import pandas as pd

# Load datasets
df_all_agree = pd.read_csv('FinancialPhraseBank-v1.0/Sentences_AllAgree.txt', sep='@', header=None, names=['sentence', 'sentiment'], encoding='ISO-8859-1')
df_75_agree = pd.read_csv('FinancialPhraseBank-v1.0/Sentences_75Agree.txt', sep='@', header=None, names=['sentence', 'sentiment'], encoding='ISO-8859-1')
df_66_agree = pd.read_csv('FinancialPhraseBank-v1.0/Sentences_66Agree.txt', sep='@', header=None, names=['sentence', 'sentiment'], encoding='ISO-8859-1')
df_50_agree = pd.read_csv('FinancialPhraseBank-v1.0/Sentences_50Agree.txt', sep='@', header=None, names=['sentence', 'sentiment'], encoding='ISO-8859-1')

# Choose one of the datasets for the project
df = df_all_agree  # Using the dataset with 100% agreement
# Display the first few rows to confirm it loaded correctly
print(df.head())


                                            sentence sentiment
0  According to Gran , the company has no plans t...   neutral
1  For the last quarter of 2010 , Componenta 's n...  positive
2  In the third quarter of 2010 , net sales incre...  positive
3  Operating profit rose to EUR 13.1 mn from EUR ...  positive
4  Operating profit totalled EUR 21.1 mn , up fro...  positive


In [7]:
from transformers import BertTokenizer, BertForSequenceClassification, pipeline

# Load the model and tokenizer
model = BertForSequenceClassification.from_pretrained("ahmedrachid/FinancialBERT-Sentiment-Analysis", num_labels=3)
tokenizer = BertTokenizer.from_pretrained("ahmedrachid/FinancialBERT-Sentiment-Analysis")

# Initialize the sentiment analysis pipeline
nlp = pipeline("sentiment-analysis", model=model, tokenizer=tokenizer)

# Example sentences
sentences = [
    "Operating profit rose to EUR 13.1 mn from EUR 8.7 mn in the corresponding period in 2007 representing 7.7 % of net sales.",
    "Bids or offers include at least 1,000 shares and the value of the shares must correspond to at least EUR 4,000.",
    "Raute reported a loss per share of EUR 0.86 for the first half of 2009, against EPS of EUR 0.74 in the corresponding period of 2008."
]

# Analyze the sentiment
results = nlp(sentences)
print(results)


C:\Users\kmusodza\AppData\Local\Programs\Python\Python312\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
C:\Users\kmusodza\AppData\Local\Programs\Python\Python312\Lib\site-packages\huggingface_hub\file_download.py:148: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\kmusodza\.cache\huggingface\hub\models--ahmedrachid--FinancialBERT-Sentiment-Analysis. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Develope

[{'label': 'positive', 'score': 0.9998133778572083}, {'label': 'neutral', 'score': 0.9997822642326355}, {'label': 'negative', 'score': 0.9877365231513977}]


### Clean the Data
We'll clean the data by removing any unnecessary characters and whitespace.

In [8]:
import re

def clean_text(text):
    # Remove non-alphabetic characters and convert to lowercase
    text = re.sub(r'[^a-zA-Z\s]', '', text)
    text = text.lower().strip()
    return text

df['sentence'] = df['sentence'].apply(clean_text)
df


sentence sentiment
0     according to gran  the company has no plans to...   neutral
1     for the last quarter of   componenta s net sal...  positive
2     in the third quarter of   net sales increased ...  positive
3     operating profit rose to eur  mn from eur  mn ...  positive
4     operating profit totalled eur  mn  up from eur...  positive
...                                                 ...       ...
2259  operating result for the month period decrease...  negative
2260  helsinki thomson financial  shares in cargotec...  negative
2261  london marketwatch  share prices ended lower i...  negative
2262  operating profit fell to eur  mn from eur  mn ...  negative
2263  sales in finland decreased by   in january  wh...  negative

[2264 rows x 2 columns]

### Tokenization
We will tokenize the sentences using NLTK.

In [9]:
import nltk
nltk.download('punkt')
from nltk.tokenize import word_tokenize

df['tokens'] = df['sentence'].apply(word_tokenize)


[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\kmusodza\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping tokenizers\punkt.zip.


### Lemmatization
Lemmatize the tokens to convert them to their base forms.

In [10]:
nltk.download('wordnet')
from nltk.stem import WordNetLemmatizer

lemmatizer = WordNetLemmatizer()

def lemmatize_tokens(tokens):
    return [lemmatizer.lemmatize(token) for token in tokens]

df['tokens'] = df['tokens'].apply(lemmatize_tokens)


[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\kmusodza\AppData\Roaming\nltk_data...


### Vectorization
Convert the text data into numerical format using TF-IDF for the Naive Bayes model and word embeddings for the BERT model.

*TF-IDF Vectorization for Naive Bayes*

In [11]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Convert tokens back to sentences for TF-IDF vectorization
df['cleaned_sentence'] = df['tokens'].apply(lambda x: ' '.join(x))

# TF-IDF vectorization
vectorizer = TfidfVectorizer(max_features=5000)
X_tfidf = vectorizer.fit_transform(df['cleaned_sentence'])


#### Word Embeddings for BERT

In [12]:
from transformers import BertTokenizer

# Load BERT tokenizer
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased')

# Tokenize sentences for BERT
df['bert_tokens'] = df['sentence'].apply(lambda x: tokenizer.encode(x, add_special_tokens=True))
df.head(8)


C:\Users\kmusodza\AppData\Local\Programs\Python\Python312\Lib\site-packages\huggingface_hub\file_download.py:148: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\kmusodza\.cache\huggingface\hub\models--bert-base-uncased. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to see activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


sentence sentiment  \
0  according to gran  the company has no plans to...   neutral   
1  for the last quarter of   componenta s net sal...  positive   
2  in the third quarter of   net sales increased ...  positive   
3  operating profit rose to eur  mn from eur  mn ...  positive   
4  operating profit totalled eur  mn  up from eur...  positive   
5  finnish talentum reports its operating profit ...  positive   
6  clothing retail chain seppl s sales increased ...  positive   
7  consolidated net sales increased   to reach eu...  positive   

                                              tokens  \
0  [according, to, gran, the, company, ha, no, pl...   
1  [for, the, last, quarter, of, componenta, s, n...   
2  [in, the, third, quarter, of, net, sale, incre...   
3  [operating, profit, rose, to, eur, mn, from, e...   
4  [operating, profit, totalled, eur, mn, up, fro...   
5  [finnish, talentum, report, it, operating, pro...   
6  [clothing, retail, chain, seppl, s, sale, incr...   
7  [consolidated, net, sale, increased, to, reach...   

                                    cleaned_sentence  \
0  according to gran the company ha no plan to mo...   
1  for the last quarter of componenta s net sale ...   
2  in the third quarter of net sale increased by ...   
3  operating profit rose to eur mn from eur mn in...   
4  operating profit totalled eur mn up from eur m...   
5  finnish talentum report it operating profit in...   
6  clothing retail chain seppl s sale increased b...   
7  consolidated net sale increased to reach eur m...   

                                         bert_tokens  
0  [101, 2429, 2000, 12604, 1996, 2194, 2038, 205...  
1  [101, 2005, 1996, 2197, 4284, 1997, 6922, 2050...  
2  [101, 1999, 1996, 2353, 4284, 1997, 5658, 4341...  
3  [101, 4082, 5618, 3123, 2000, 7327, 2099, 2409...  
4  [101, 4082, 5618, 2561, 3709, 7327, 2099, 2409...  
5  [101, 6983, 5848, 2819, 4311, 2049, 4082, 5618...  
6  [101, 5929, 7027, 4677, 19802, 24759, 1055, 43...  
7  [101, 10495, 5658, 4341, 3445, 2000, 3362, 732...

## Model Implementation
We'll implement and evaluate two models: a Naive Bayes classifier using TF-IDF vectors and a BERT model for sentiment analysis.

### Implement the Naive Bayes Classifier

- Split data
- Train data classifier
- Predict on test set
- Evaluate classsifer

In [13]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Map sentiments to numerical labels
sentiment_mapping = {'positive': 1, 'neutral': 0, 'negative': -1}
df['sentiment_label'] = df['sentiment'].map(sentiment_mapping)

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_tfidf, df['sentiment_label'], test_size=0.2, random_state=42)

# Train the Naive Bayes classifier
nb_classifier = MultinomialNB()
nb_classifier.fit(X_train, y_train)

# Predict on the test set
y_pred = nb_classifier.predict(X_test)

# Evaluate the Naive Bayes classifier
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='macro')
recall = recall_score(y_test, y_pred, average='macro')
f1 = f1_score(y_test, y_pred, average='macro')

# Print evaluation metrics
print(f"Naive Bayes Classifier\nAccuracy: {accuracy}\nPrecision: {precision}\nRecall: {recall}\nF1-Score: {f1}")


Naive Bayes Classifier
Accuracy: 0.7858719646799117
Precision: 0.8327235772357723
Recall: 0.5789893571514125
F1-Score: 0.5767244438177758


### BERT Model for Sentiment Analysis
#### Implement the BERT Model

We'll use the transformers library to fine-tune a pre-trained BERT model.

In [14]:
import torch
from torch.utils.data import DataLoader, Dataset
from transformers import BertForSequenceClassification, AdamW, get_linear_schedule_with_warmup
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split

# Dataset class for BERT
class FinancialSentimentDataset(Dataset):
    def __init__(self, sentences, labels, tokenizer, max_len):
        self.sentences = sentences
        self.labels = labels
        self.tokenizer = tokenizer
        self.max_len = max_len
    
    def __len__(self):
        return len(self.sentences)
    
    def __getitem__(self, item):
        sentence = str(self.sentences[item])
        label = self.labels[item]
        
        encoding = self.tokenizer.encode_plus(
            sentence,
            add_special_tokens=True,
            max_length=self.max_len,
            return_token_type_ids=False,
            padding='max_length',
            truncation=True,
            return_attention_mask=True,
            return_tensors='pt'
        )
        
        return {
            'sentence_text': sentence,
            'input_ids': encoding['input_ids'].flatten(),
            'attention_mask': encoding['attention_mask'].flatten(),
            'labels': torch.tensor(label, dtype=torch.long)
        }

# Define the device
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Encode labels
le = LabelEncoder()
df['label'] = le.fit_transform(df['sentiment'])

# Create dataset and dataloader
MAX_LEN = 128
BATCH_SIZE = 16

train_sentences, test_sentences, train_labels, test_labels = train_test_split(df['sentence'], df['label'], test_size=0.2, random_state=42)

train_dataset = FinancialSentimentDataset(train_sentences.values, train_labels.values, tokenizer, MAX_LEN)
test_dataset = FinancialSentimentDataset(test_sentences.values, test_labels.values, tokenizer, MAX_LEN)

train_dataloader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True)
test_dataloader = DataLoader(test_dataset, batch_size=BATCH_SIZE, shuffle=False)

# Load pre-trained BERT model for sequence classification
model = BertForSequenceClassification.from_pretrained("bert-base-uncased", num_labels=3)
model = model.to(device)

# Set up optimizer and scheduler
optimizer = AdamW(model.parameters(), lr=2e-5, correct_bias=False)
total_steps = len(train_dataloader) * 4  # 4 epochs
scheduler = get_linear_schedule_with_warmup(optimizer, num_warmup_steps=0, num_training_steps=total_steps)

# Training loop
def train_epoch(model, dataloader, optimizer, scheduler, device):
    model = model.train()
    total_loss = 0
    correct_predictions = 0
    
    for batch in dataloader:
        input_ids = batch['input_ids'].to(device)
        attention_mask = batch['attention_mask'].to(device)
        labels = batch['labels'].to(device)
        
        outputs = model(input_ids=input_ids, attention_mask=attention_mask, labels=labels)
        loss = outputs.loss
        logits = outputs.logits
        
        _, preds = torch.max(logits, dim=1)
        correct_predictions += torch.sum(preds == labels)
        total_loss += loss.item()
        
        loss.backward()
        optimizer.step()
        scheduler.step()
        optimizer.zero_grad()
    
    return correct_predictions.double() / len(dataloader.dataset), total_loss / len(dataloader)

def eval_model(model, dataloader, device):
    model = model.eval()
    correct_predictions = 0
    total_loss = 0
    
    with torch.no_grad():
        for batch in dataloader:
            input_ids = batch['input_ids'].to(device)
            attention_mask = batch['attention_mask'].to(device)
            labels = batch['labels'].to(device)
            
            outputs = model(input_ids=input_ids, attention_mask=attention_mask, labels=labels)
            loss = outputs.loss
            logits = outputs.logits
            
            _, preds = torch.max(logits, dim=1)
            correct_predictions += torch.sum(preds == labels)
            total_loss += loss.item()
    
    return correct_predictions.double() / len(dataloader.dataset), total_loss / len(dataloader)

# Train the BERT model
EPOCHS = 4

for epoch in range(EPOCHS):
    print(f"Epoch {epoch + 1}/{EPOCHS}")
    train_acc, train_loss = train_epoch(model, train_dataloader, optimizer, scheduler, device)
    print(f"Train loss {train_loss} accuracy {train_acc}")
    
    val_acc, val_loss = eval_model(model, test_dataloader, device)
    print(f"Val loss {val_loss} accuracy {val_acc}")


Some weights of BertForSequenceClassification were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
C:\Users\kmusodza\AppData\Local\Programs\Python\Python312\Lib\site-packages\transformers\optimization.py:457: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(


Epoch 1/4
Train loss 0.36738332534176216 accuracy 0.8586416344561016
Val loss 0.1509924430785508 accuracy 0.9492273730684326
Epoch 2/4
Train loss 0.09417342842278774 accuracy 0.9745996686913307
Val loss 0.14169721671476446 accuracy 0.9580573951434879
Epoch 3/4
Train loss 0.044061490187519474 accuracy 0.9895085588072888
Val loss 0.1489185376959885 accuracy 0.9602649006622517
Epoch 4/4
Train loss 0.02463073427421286 accuracy 0.992821645499724
Val loss 0.15294025441760134 accuracy 0.9558498896247241


## Model Evaluation

### Naive Bayes Classifier Evaluation
- Use Confusion matrix to viusally evaluate the model

In [15]:
from sklearn.metrics import classification_report

# Predict on the test set
y_pred = nb_classifier.predict(X_test)

# Evaluate the Naive Bayes classifier
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='macro')
recall = recall_score(y_test, y_pred, average='macro')
f1 = f1_score(y_test, y_pred, average='macro')

# Print evaluation metrics
print("Naive Bayes Classifier Evaluation")
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-Score: {f1:.4f}")
print("\nClassification Report:\n", classification_report(y_test, y_pred))


Naive Bayes Classifier Evaluation
Accuracy: 0.7859
Precision: 0.8327
Recall: 0.5790
F1-Score: 0.5767

Classification Report:
               precision    recall  f1-score   support

          -1       1.00      0.09      0.16        56
           0       0.82      0.98      0.89       276
           1       0.68      0.67      0.67       121

    accuracy                           0.79       453
   macro avg       0.83      0.58      0.58       453
weighted avg       0.81      0.79      0.74       453



In [16]:
from sklearn.metrics import classification_report, confusion_matrix
import seaborn as sns
import matplotlib.pyplot as plt

# Predict on test set
y_pred = nb_classifier.predict(X_test)

# Evaluate model
print("Naive Bayes Classifier Evaluation")
print(classification_report(y_test, y_pred))

# Confusion matrix
nb_cm = confusion_matrix(y_test, y_pred)

plt.figure(figsize=(5,4))
sns.heatmap(nb_cm, annot=True, fmt='d', cmap='Blues', xticklabels=['Negative', 'Neutral', 'Positive'], yticklabels=['Negative', 'Neutral', 'Positive'])
plt.title('Confusion Matrix - Naive Bayes')
plt.xlabel('Predicted')
plt.ylabel('Actual')
plt.show()


ModuleNotFoundError: No module named 'seaborn'

### BERT Model Evaluation

- We'll use the evaluation function defined earlier to calculate the metrics for the BERT model.
- Use Confusion matrix to viusally evaluate the model

In [ ]:
def evaluate_model(model, dataloader, device):
    model.eval()
    predictions = []
    true_labels = []
    
    with torch.no_grad():
        for batch in dataloader:
            input_ids = batch['input_ids'].to(device)
            attention_mask = batch['attention_mask'].to(device)
            labels = batch['labels'].to(device)
            
            outputs = model(input_ids=input_ids, attention_mask=attention_mask)
            logits = outputs.logits
            preds = torch.argmax(logits, dim=1).flatten()
            
            predictions.extend(preds)
            true_labels.extend(labels)
    
    accuracy = accuracy_score(true_labels, predictions)
    precision = precision_score(true_labels, predictions, average='macro')
    recall = recall_score(true_labels, predictions, average='macro')
    f1 = f1_score(true_labels, predictions, average='macro')
    
    return accuracy, precision, recall, f1, true_labels, predictions

# Evaluate the BERT model
bert_accuracy, bert_precision, bert_recall, bert_f1, bert_true, bert_pred = evaluate_model(model, test_dataloader, device)

print("BERT Model Evaluation")
print(f"Accuracy: {bert_accuracy:.4f}")
print(f"Precision: {bert_precision:.4f}")
print(f"Recall: {bert_recall:.4f}")
print(f"F1-Score: {bert_f1:.4f}")
print("\nClassification Report:\n", classification_report(bert_true, bert_pred))


In [ ]:
# Evaluate the BERT model on the test set
bert_accuracy, bert_precision, bert_recall, bert_f1, bert_true, bert_pred = evaluate_model(model, test_dataloader, device)

print("BERT Model Evaluation")
print(f"Accuracy: {bert_accuracy:.4f}")
print(f"Precision: {bert_precision:.4f}")
print(f"Recall: {bert_recall:.4f}")
print(f"F1-Score: {bert_f1:.4f}")
print("\nClassification Report:\n", classification_report(bert_true, bert_pred))

# Confusion matrix
bert_cm = confusion_matrix(bert_true, bert_pred)

plt.figure(figsize=(5,4))
sns.heatmap(bert_cm, annot=True, fmt='d', cmap='Blues', xticklabels=['Negative', 'Neutral', 'Positive'], yticklabels=['Negative', 'Neutral', 'Positive'])
plt.title('Confusion Matrix - BERT')
plt.xlabel('Predicted')
plt.ylabel('Actual')
plt.show()


### Save Trained Models

In [ ]:
import joblib

# Save Naive Bayes model
joblib.dump(nb_classifier, 'naive_bayes_model.pkl')
# Save TF-IDF vectorizer
joblib.dump(vectorizer, 'tfidf_vectorizer.pkl')

# Save the trained BERT model and tokenizer
model.save_pretrained('bert_model')
tokenizer.save_pretrained('bert_tokenizer')

## Comparison and Analysis of Naive Bayes and BERT Models
| Metric          | Naive Bayes | BERT      |
|-----------------|-------------|-----------|
| Accuracy        | 0.7859      | 0.9669    |
| Precision       | 0.8327      | 0.9538    |
| Recall          | 0.5790      | 0.9553    |
| F1-Score        | 0.5767      | 0.9545    |


Classification Reports

Naive Bayes Classifier:

            precision    recall  f1-score   support

          -1       1.00      0.09      0.16        56
           0       0.82      0.98      0.89       276
           1       0.68      0.67      0.67       121

    accuracy                           0.79       453
   macro avg       0.83      0.58      0.58       453
weighted avg       0.81      0.79      0.74       453


BERT Model:

               precision    recall  f1-score   support

           0       0.93      0.95      0.94        56
           1       0.98      0.99      0.98       276
           2       0.95      0.93      0.94       121

    accuracy                           0.97       453
   macro avg       0.95      0.96      0.95       453
weighted avg       0.97      0.97      0.97       453


### **Analysis**

- Naive Bayes: Simple and fast, but struggles with negative sentiments.

- BERT: Highly accurate and robust, capable of understanding complex language patterns and context.


# Conclusion

### Naive Bayes Classifier:

**Strengths**:

- Simplicity and speed: The Naive Bayes classifier is easy to implement and computationally efficient.
- High precision for neutral sentiments (0.82) and acceptable accuracy (0.7859).

**Weaknesses**:

- Poor recall for negative sentiments (0.09), indicating that the model struggles to identify negative instances.
- Lower overall F1-Score (0.5767), especially for negative sentiments, suggesting that the model doesn't perform well across all classes.


### Bert model
**Strengths**:

- High accuracy (0.9669), indicating that the model is very good at predicting the correct sentiment.
- Balanced precision (0.9538) and recall (0.9553), showing that the model can reliably identify all sentiment classes.
- High F1-Score (0.9545) across all classes, demonstrating the model's robustness and effectiveness in financial sentiment analysis.

**Weaknesses**:

- Computationally intensive: Training and inference with BERT require significant resources.
- Complexity: Implementing and fine-tuning BERT is more complex compared to simpler models like Naive Bayes.


**Conclusion**:
- The BERT model is highly effective for financial sentiment analysis, providing accurate and reliable predictions.
- The Naive Bayes classifier, although simpler, is less effective in this context but could be useful for quick, preliminary analysis.
- Future work could explore further fine-tuning of BERT, experimenting with other advanced models, or combining multiple models for even better performance.